src/models/kernels.hh

```diff
@@ -95,7 +94,7 @@ namespace rascal {
                     representation_name)};
 
             kernel(ii_A, ii_B) =
-                propA.dot(propB).unaryExpr(integer_power).mean();
+                propA.dot(propB).array().pow(this->zeta).mean();
```

Comment:
Eigen pow is also built-in for integer types: https://eigen.tuxfamily.org/dox/group__CoeffwiseMathFunctions.html

Reply:
unfortunately not. it works with integer / integer pow which is not the case here so it cast the int to double. Please put back the intended behavior (integer division).

Reply:
> Please put back the intended behavior (integer power).

I don't see what behavior change you are talking about here

Reply:
If you are worried about a potential performance loss (which I would like to see demonstrated), we can also use [square](https://eigen.tuxfamily.org/dox/classEigen_1_1ArrayBase.html#a4fe59f7f07a8b445e47b0a6d9f1c15a1) and `cube` functions.

Reply:
There is no question about the efficiency difference between exp + log compared to a few multiplications. 
The previous behaviour is to make sure integer power is used (which is not the case anymore).

Reply:
> There is no question about the efficiency difference between exp + log compared to a few multiplications.

std::pow is usually not implemented as exp + log(see the implementation of musl here https://git.musl-libc.org/cgit/musl/tree/src/math/pow.c), and not necessary slower than repeated multiplication: https://martin-ueding.de/articles/efficiency-of-pow-function/index.html.

When talking about performances, better to measure twice. Here, since `unaryExpr` is opaque to Eigen optimisations, this might also be the case, but again should be benchmarked (I'll set something up tomorrow).

And citing Knuth, "premature optimization is the root of all evil": I'd rather optimize for readability and further modifications than gut-feeling performances. This way, when I want to to optimize for performances, I can more easily see what is going on and check that my changes had the intended effects.

Reply:
your link refers to a comparison for power 2 which is often a special case in the pow implementation and the gains are compiler dependents. 
Before removing the usage of std::pow in spherical expansion it used to take up to 20% of runtime (benchmarked with vtune) and after the integer pow function was less than 2% of runtime. 
I agree this is not a polished benchmark but to me this is good enough to warrant not using std::pow when possible.

Reply:
Here is a micro benchmark, where the power is given as a command line argument to make it invisible to the optimizer.: https://gist.github.com/Luthaf/6170745dd6eb6654699dbb033e5ac6a8

The TL;DR is `Eigen::pow == std::pow > functor == explicit for loop > Eigen::square/Eigen::cube`

If performance is really an issue, then the best solution is to explicitly use `square/cube`, and fallback to an explicit for-loop for higher values of zeta.

But again, I am not convinced that this operation is worth optimizing too much: I think the memory accesses to get the properties would take much more time that this (would have to be benchmarked again).

Reply:
it makes sense since Eigen::square/Eigen::cube have sse2/avx instructions. the functor is to avoid the for loop and if the performance are important it can be written with vectorized instructions.